driver/sliding_expiration_cache.cc

```diff
@@ -0,0 +1,136 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// This program is free software; you can redistribute it and/or modify
+// it under the terms of the GNU General Public License, version 2.0
+// (GPLv2), as published by the Free Software Foundation, with the
+// following additional permissions:
+//
+// This program is distributed with certain software that is licensed
+// under separate terms, as designated in a particular file or component
+// or in the license documentation. Without limiting your rights under
+// the GPLv2, the authors of this program hereby grant you an additional
+// permission to link the program and your derivative works with the
+// separately licensed software that they have included with the program.
+//
+// Without limiting the foregoing grant of rights under the GPLv2 and
+// additional permission as to separately licensed software, this
+// program is also subject to the Universal FOSS Exception, version 1.0,
+// a copy of which can be found along with its FAQ at
+// http://oss.oracle.com/licenses/universal-foss-exception.
+//
+// This program is distributed in the hope that it will be useful, but
+// WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+// See the GNU General Public License, version 2.0, for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see
+// http://www.gnu.org/licenses/gpl-2.0.html.
+
+#include "sliding_expiration_cache.h"
+
+#include <utility>
+#include <vector>
+
+template <class K, class V>
+void SLIDING_EXPIRATION_CACHE<K, V>::remove_and_dispose(K key) {
+  if (this->cache.count(key)) {
+    CACHE_ITEM* cache_item = this->cache[key];
+    if (item_disposal_func != nullptr) {
+      item_disposal_func->dispose(cache_item->item);
+    }
+    this->cache.erase(key);
+  }
+}
+
+template <class K, class V>
+void SLIDING_EXPIRATION_CACHE<K, V>::clean_up() {
+  if (this->clean_up_time_nanos.load() > std::chrono::steady_clock::now()) {
+    return;
+  }
+
+  this->clean_up_time_nanos =
+      std::chrono::steady_clock::now() + std::chrono::nanoseconds(this->clean_up_interval_nanos);
+
+  std::vector<K> keys;
+  keys.reserve(this->cache.size());
+  for (auto& [key, cache_item] : this->cache) {
```

Comment:
Storing keys into a separate vector to avoid removing items from the map while iterating over it.